hazelcast/src/test/java/com/hazelcast/map/MapTransactionLockingTest.java

```diff
@@ -0,0 +1,450 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IMap;
+import com.hazelcast.core.TransactionalMap;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.TestHazelcastInstanceFactory;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import com.hazelcast.transaction.TransactionException;
+import com.hazelcast.transaction.TransactionalTask;
+import com.hazelcast.transaction.TransactionalTaskContext;
+import java.util.concurrent.CountDownLatch;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class MapTransactionLockingTest extends HazelcastTestSupport {
+
+    @Test
+    public void testTxnReplace_whenReplaceIfSameFails_keyShouldRemainUnlocked() throws InterruptedException {
+        final String mapName = randomMapName();
+        final String key = "k";
+        final String value = "v";
+        final TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(1);
+        final HazelcastInstance hazelcastInstance = factory.newHazelcastInstance(getConfig());
+
+        IMap<String, Object> map = hazelcastInstance.getMap(mapName);
+        map.put(key, value);
+
+        hazelcastInstance.executeTransaction(new TransactionalTask<Object>() {
+            @Override
+            public Object execute(TransactionalTaskContext transactionContext) throws TransactionException {
+                TransactionalMap<String, Object> map = transactionContext.getMap(mapName);
+                map.replace(key, value + "other", value);
+                return null;
+            }
+        });
+        assertFalse("Key remains locked!", map.isLocked(key));
+    }
+
+    @Test
+    public void testTxnReplace_whenReplaceIfSameFails_keyShouldRemainUnlockedDuringTransaction() throws InterruptedException {
+        final String mapName = randomMapName();
+        final String key = "k";
+        final String value = "v";
+        final TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(1);
+        final HazelcastInstance hazelcastInstance = factory.newHazelcastInstance(getConfig());
+        IMap<String, Object> map = hazelcastInstance.getMap(mapName);
+
+        final CountDownLatch startAssertion = new CountDownLatch(1);
+        final CountDownLatch finishAssertion = new CountDownLatch(1);
+        final CountDownLatch finishTransaction = new CountDownLatch(1);
+
+        new Thread(new Runnable() {
```

Comment:
no need to introduce a thread here, you can use `map.isLocked(key)` after `map.replace`
this will dramatically decrease the test complexity